cirq/contrib/quirk/cells/arithmetic_cells.py

```diff
@@ -0,0 +1,308 @@
+# Copyright 2019 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import inspect
+from typing import (Callable, Optional, Union, Iterable, Sequence, Iterator,
+                    Tuple, Any, cast, List)
+
+import cirq
+from cirq import ops, value
+from cirq.contrib.quirk.cells.cell import Cell, CellMaker, CELL_SIZES
+
+
+@value.value_equality
+class QuirkArithmeticOperation(ops.ArithmeticOperation):
+    """Applies an arithmetic to a target and some inputs.
+
+    Uses quirk-specific implicit effects like assuming that the presence of an
+    'r' input implies modular arithmetic (meaning that e.g. values larger than
+    the modulus will not be affected).
+    """
+
+    def __init__(self, identifier: str,
+                 operation: 'cirq.contrib.quirk.QuirkArithmeticLambda',
+                 target: Sequence['cirq.Qid'],
+                 inputs: Sequence[Optional[Union[Sequence['cirq.Qid'], int]]]):
+        """
+        Args:
+            identifier: The quirk identifier string for this operation.
+            operation: A repr-able lambda that determines the new value of the
+                target based on its current value and the values of the inputs.
+            target: The target qubit register.
+            inputs: Qubit registers (or classical constants) that
+                determine what happens to the target.
+        """
+        if operation.is_modular:
+            r = cast(Union[Sequence['cirq.Qid'], int], inputs[-1])
```

Comment:
That used to be the way it worked. It made things more complicated, because this parameter had to be fed through several places in the code. For example, it ended up being in the repr shown to users. The specialization of the meaning of r is a quirk convention that it makes sense to reflect in the implementation.

Reply:
Makes sense.